internal/selfupdate/download.go

```diff
@@ -74,33 +66,47 @@ func extractToFile(buf []byte, filename, target string, printf func(string, ...i
 		rd = file
 	}
 
-	err = os.Remove(target)
-	if os.IsNotExist(err) {
-		err = nil
-	}
+	// Write everything to a temp file
+	new, err := os.CreateTemp("", "restic")
 	if err != nil {
-		return fmt.Errorf("unable to remove target file: %v", err)
+		return err
 	}
 
-	dest, err := os.OpenFile(target, os.O_CREATE|os.O_EXCL|os.O_WRONLY, mode)
+	n, err := io.Copy(new, rd)
 	if err != nil {
+		new.Close()
+		os.Remove(new.Name())
 		return err
 	}
 
-	n, err := io.Copy(dest, rd)
-	if err != nil {
-		_ = dest.Close()
-		_ = os.Remove(dest.Name())
+	if err = new.Close(); err != nil {
 		return err
 	}
 
-	err = dest.Close()
-	if err != nil {
+	printf("saved %d bytes in %v\n", n, new.Name())
+
+	mode := os.FileMode(0755)
+	// attempt to find the original mode
+	if fi, err := os.Lstat(target); err == nil {
+		mode = fi.Mode()
+	}
+
+	// Rename (rather than remove) the running version. The running binary will be locked
+	// on some platforms (Windows) and cannot be removed while executing.
+	backup := filepath.Join(os.TempDir(), filepath.Base(target)+".bak")
+	if _, err = os.Stat(backup); err == nil {
+		os.Remove(backup)
+	}
+	if err = os.Rename(target, backup); err != nil {
+		return fmt.Errorf("unable to rename target file: %v", err)
+	}
+
+	// Rename the temp file to the final location atomically
```

Comment:
This isn't fully atomic, because we've written the new file to the tempdir, which may be in a different mount point (on Unix) or drive (on Windows). To get an atomic move, we'd have to put the new file in the same directory:

```
new, err := os.CreateTemp(filepath.Dir(target), "restic")
```

For the same reason, the backup should be in the same directory as well. That also prevents leaking information when the target dir is encrypted but the tempdir is not.

Reply:
Done. I'll note that the main tradeoff here is that we don't get any automatic cleanup of these temp files in this case. Should the operation fail for some reason, the temp file could theoretically be orphaned in the working directory... Although this should hopefully be a very rare occurrence.